extensions/amp-video/0.1/video-cache.js

```diff
@@ -95,3 +103,34 @@ function applySourcesToVideo(videoEl, sources) {
       videoEl.insertBefore(sourceEl, videoEl.firstChild);
     });
 }
+
+/**
+ * Moves the src attribute to a source element to enable playing from multiple
+ * sources: the cached ones and the fallback initial src.
+ * @param {!Element} videoEl
+ * @param {!Window} win
+ * @return {!Promise}
+ */
+function replaceSrcWithSourceElement(videoEl, win) {
+  if (!videoEl.hasAttribute('src')) {
+    return;
+  }
+  const sourceEl = win.document.createElement('source');
+  const srcAttr = videoEl.getAttribute('src');
+  sourceEl.setAttribute('src', srcAttr);
+
+  const typeAttr = videoEl.getAttribute('type');
+  if (typeAttr) {
+    sourceEl.setAttribute('type', typeAttr);
+  }
+
+  // Remove the src attr so the source children can play.
+  videoEl.removeAttribute('src');
+
+  // Remove all existing sources as they are never supposed to play for a video
+  // that has a src, cf https://html.spec.whatwg.org/#concept-media-load-algorithm
+  const sourceEls = toArray(videoEl.querySelectorAll('source'));
+  sourceEls.forEach((el) => removeElement(el));
```

Comment:
There's `iterateCursor` which does this without creating the array first.

```suggestion
  const sourceEls = videoEl.querySelectorAll('source');
  iterateCursor(sourceEls, (el) => removeElement(el));
```